File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> mpp = new HashMap<>();
        int l = 0, ans = 0;

        for (int r = 0; r < s.length(); r++) {
            char ch = s.charAt(r);

            if (mpp.containsKey(ch) && mpp.get(ch) >= l) {
                l = mpp.get(ch) + 1;
            }

            mpp.put(ch, r);
            ans = Math.max(ans, r - l + 1);
        }

        return ans;
    }
}

File: Data Structures & Algorithms/interleaving-string/submission-0.cpp
class Solution {
   public:
    vector<vector<int>> dp;

    bool isInterleave(string s1, string s2, string s3) {
        if (s1.size() + s2.size() != s3.size()) return false;

        dp.assign(s1.size() + 1, vector<int>(s2.size() + 1, -1));

        return dfs(0, 0, s1, s2, s3);
    }

    bool dfs(int i, int j, string& s1, string& s2, string& s3) {
        if (i == s1.size() && j == s2.size()) return true;

        if (dp[i][j] != -1) return dp[i][j];

        int k = i + j;

        bool ans = false;

        if (i < s1.size() && s1[i] == s3[k]) {
            ans |= dfs(i + 1, j, s1, s2, s3);
        }

        if (!ans && j < s2.size() && s2[j] == s3[k]) {
            ans |= dfs(i, j + 1, s1, s2, s3);
        }

        return dp[i][j] = ans;
    }
};